default/scripting/common/enqueue.macros

```diff
@@ -30,7 +30,10 @@ LOCATION_ALLOW_BUILD_IF_PREREQ_ENQUEUED
 
 ENQUEUE_BUILD_ONE_PER_PLANET
 '''And [
-        Not Contains Building name = CurrentContent
+        // Only check for visible buildings. Own buildings will always be, but the client may have seen a building once,
+        // which has been destroyed unbeknown to the client. Then this building will forever remain in the client's
+        // context, but it should not stop a player from rebuilding the same type.
+        Number high = 0 condition = And [ Building name = CurrentContent ContainedBy Source VisibleToEmpire empire = Source.Owner ]
```

Comment:
IMO this part does not matter. two lines below it checks if the enqueued building is owned by the building empire.
so it does not matter if there are ghosts of other empires or not.

Reply:
The conditions here are all for the enqueue location, i.e. the planet the building should be built at.
I cannot check the owner of a building in this context, that's why I used the Number expression.

Reply:
Rather than `Number ... condition = ... ContainedBy Source` you could probably use a `Contains` condition applied to the enqueue location

Reply:
If there is a ghost building at planet A and a real one at planet B, the player should still be able to enqueue one at A, that's why I specifically have to check that there is a building at the enqueue location and **that building** is visible / belongs to the player.

Perhaps adding Contains before the number expression makes the overall condition more efficient, because the Number expression would be skipped in most cases, but I do need the ContainsBy.

Reply:
> The conditions here are all for the enqueue location, i.e. the planet the building should be built at.
> I cannot check the owner of a building in this context, that's why I used the Number expression.

At least one of us is confused here. If there are buildings on the planet they have an owner.
also the planet has an owner. 
The owner of the planet should not be able to enqueue a building if the planet contains the same building if it is owned by them. So something like this:
```
 Not Contains And [ Building name = "BLD_COL_{name}" OwnedBy empire = Source.Owner ]
```

Reply:
Thank you for the example, I simply did not see how to do it that way. That is of course much better.